src/webui/api/appcontroller.cpp

```diff
@@ -1312,6 +1357,75 @@ void AppController::setCookiesAction()
     setResult(QString());
 }
 
+void AppController::clientDataAction()
+{
+    const QString keysParam {params()[u"keys"_s]};
+    if (keysParam.isEmpty())
+    {
+        setResult(m_clientData);
+        return;
+    }
+
+    QJsonParseError jsonError;
+    const auto keysJsonDocument = QJsonDocument::fromJson(keysParam.toUtf8(), &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+        throw APIError(APIErrorType::BadParams, jsonError.errorString());
+    if (!keysJsonDocument.isArray())
+        throw APIError(APIErrorType::BadParams, tr("keys must be an array"));
+
+    QJsonObject clientData;
+    for (const QJsonValue &keysJsonVal : asConst(keysJsonDocument.array()))
+    {
+        if (!keysJsonVal.isString())
+            throw APIError(APIErrorType::BadParams, tr("key must be a string"));
+
+        const QString &key = keysJsonVal.toString();
+        if (const auto iter = m_clientData.constFind(key); iter != m_clientData.constEnd())
+            clientData.insert(key, iter.value());
```

Comment:
Wouldn't the following simplify things?
```suggestion
        clientData.insert(key, m_clientData[key]);
```

Reply:
This changes the behavior of the API. Currently, when a nonexistent key is requested, it's omitted from the response. With your change, we now include the key with a `null` value.

Reply:
> With your change, we now include the key with a `null` value.

Are you sure about this?
`m_clientData[key]` returns `QJsonValue::Undefined` in case when `key` doesn't exist.
`QJsonValue::insert()` doesn't insert anything in case when `value` is `QJsonValue::Undefined` (it is even remove existing `key`).

Reply:
Yes, see test results below.

Before:
```
$ curl "https://127.0.0.1:8080/api/v2/app/clientData" -k -w "\n" -X POST -d 'keys=["a"]'
{}
```

After:
```
$ curl "https://127.0.0.1:8080/api/v2/app/clientData" -k -w "\n" -X POST -d 'keys=["a"]'
{"a":null}
```

Reply:
> I have to double-check this.

Well, I found a mistake. It consists in the difference between the behavior of constant and non-constant `QJsonObject::operator[]` overloads (non-constant inserts an `QJsonValue::Null` item if key does not exist). So it would be correct to rewrite my suggestion as the following:
```c++
const QString &key = keysJsonVal.toString();
clientData.insert(key, m_clientData.value(key));
```

Reply:
But your code still reflects the intention more clearly, so let it stay as it is.